src/components/extractor.ts

```diff
@@ -0,0 +1,189 @@
+
+  return exportType
+    .getProperties()
+    .map((value): ExpandedProp => {
+      const declaration = extractDeclaration(value);
+      const type = checker.getTypeAtLocation(declaration);
+      return {
+        name: value.name,
+        type: stringifyType(type, checker),
+        rawType: type,
+        isOptional: isOptional(type),
+        description: getDescription(value.getDocumentationComment(checker), declaration),
+      };
+    })
+    .sort((a, b) => a.name.localeCompare(b.name));
+}
+
+export function extractFunctions(propsSymbol: ts.Symbol, checker: ts.TypeChecker) {
+  const propsName = propsSymbol.getName();
+  const namespaceDeclaration = [
+    // if we got the namespace directly
+    ...(propsSymbol.getDeclarations() ?? []),
+    // find namespace declaration from the interface with the same name
+    ...(checker.getDeclaredTypeOfSymbol(propsSymbol).getSymbol()?.getDeclarations() ?? []),
+  ].find(decl => decl.kind === ts.SyntaxKind.ModuleDeclaration);
+  const refType = unwrapNamespaceDeclaration(namespaceDeclaration)
+    .map(child => checker.getTypeAtLocation(child))
+    .find(type => (type.getSymbol() ?? type.aliasSymbol)?.getName() === 'Ref');
+
+  if (!refType) {
+    return [];
+  }
+  return refType
+    .getProperties()
+    .map((value): ExpandedProp => {
+      const declaration = extractDeclaration(value);
+      const type = checker.getTypeAtLocation(declaration);
+      const realType = type.getNonNullableType();
+      if (realType.getCallSignatures().length === 0) {
+        throw new Error(
+          `${propsName}.Ref should contain only methods, "${value.name}" has a "${stringifyType(type, checker)}" type`
+        );
+      }
+      return {
+        name: value.name,
+        type: stringifyType(realType, checker),
+        rawType: realType,
+        isOptional: isOptional(type),
+        description: getDescription(value.getDocumentationComment(checker), declaration),
+      };
+    })
+    .sort((a, b) => a.name.localeCompare(b.name));
+}
+
+export function extractExports(
+  componentName: string,
+  exportSymbols: ts.Symbol[],
+  checker: ts.TypeChecker,
+  extraExports: Record<string, Array<string>>
+) {
+  let componentSymbol;
+  let propsSymbol;
+  const unknownExports: Array<string> = [];
+  for (const exportSymbol of exportSymbols) {
+    if (exportSymbol.name === 'default') {
+      validateComponentType(componentName, exportSymbol, checker);
+      componentSymbol = exportSymbol;
+    } else if (exportSymbol.name === `${componentName}Props`) {
+      propsSymbol = exportSymbol;
+    } else if (!extraExports[componentName] || !extraExports[componentName].includes(exportSymbol.name)) {
+      unknownExports.push(exportSymbol.name);
+    }
+  }
+  // disabled until migration is complete
+  // if (unknownExports.length > 0) {
+  // throw new Error(`Unexpected exports in ${componentName}: ${unknownExports.join(', ')}`);
+  // }
```

Comment:
The idea is to fail the build if a `button/index.tsx` file exports anything other than `ButtonProps` and `Button` (as a default export).

But this will require additional work in downstream repositories. I will enable it after migration is completed.